### PR TITLE
New proposal: support for the google.api.HttpBody

### PR DESCRIPTION
**Pull request by wimspaargaren**

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -39,4 +39,4 @@ gopkg.in/resty.v1 v1.12.0 h1:CuXP0Pjfw9rOuY6EP+UvtNvt5DSqHpIxILZKT/quCZI=
 gopkg.in/resty.v1 v1.12.0/go.mod h1:mDo4pnntr5jdWRML875a/NmxYqAlA73dVijT2AXvQQo=
 gopkg.in/yaml.v2 v2.0.0-20170812160011-eb3733d160e7 h1:+t9dhfO+GNOIGJof6kPOAenx7YgrZMTdRPV+EsnPabk=
 gopkg.in/yaml.v2 v2.0.0-20170812160011-eb3733d160e7/go.mod h1:JAlM8MvJe8wmxCU4Bli9HhUf9+ttbYbLASfIpnQbh74=
-honnef.co/go/tools v0.0.0-20190102054323-c2f93a96b099/go.mod h1:rf3lG4BRIbNafJWhAfAdb/ePZxsR/4RtNHQocxwk9r4=
+honnef.co/go/tools v0.0.0-20190102054323-c2f93a96b099/go.mod h1:rf3lG4BRIbNafJWhAfAdb/ePZxsR/4RtNHQocxwk9r4=
```
Comment (johanbrandhorst) [inline go.sum]: I don't understand these two diffs

File: runtime/marshaler.go [hunk 2]
```diff
@@ -16,7 +16,7 @@ type Marshaler interface {
 	// NewEncoder returns an Encoder which writes bytes sequence into "w".
 	NewEncoder(w io.Writer) Encoder
 	// ContentType returns the Content-Type which this marshaler is responsible for.
-	ContentType() string
+	ContentType(v interface{}) string
```
Comment (johanbrandhorst) [inline runtime/marshaler.go]: I don't think we can change this without breaking backwards compatibility. Could we add a new interface with this new method and then type assert at marshalling time? That way we could support both new and old marshallers.

File: runtime/marshal_httpbodyproto.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
+package runtime
+
+import (
+	"google.golang.org/genproto/googleapis/api/httpbody"
+	"io"
```
Comment (johanbrandhorst) [inline runtime/marshal_httpbodyproto.go]: Please separate these imports (stdlib at the top)

**Comment by wimspaargaren** [comment]
I signed it!

File: go.sum (hunk 1 above)
Comment (wimspaargaren) [reply go.sum]: Me neither. I did however accidentally commit a changed mod/sum and replaced it with the original afterwards. Maybe something went wrong there.

File: runtime/marshaler.go (hunk 2 above)
Comment (wimspaargaren) [reply runtime/marshaler.go]: Changed it back and assert it at marshalling time. What do you think of this implementation?

File: go.mod [hunk 4]
```diff
@@ -13,4 +13,4 @@ require (
 	gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127 // indirect
 	gopkg.in/resty.v1 v1.12.0
 	gopkg.in/yaml.v2 v2.0.0-20170812160011-eb3733d160e7 // indirect
-)
+)
```
Comment (johanbrandhorst) [inline go.mod]: You should be able to ignore these changes with git, even though I can't tell what the difference it. Is it some whitespace issue?

File: runtime/marshaler.go (hunk 2 above)
Comment (wimspaargaren) [reply runtime/marshaler.go]: I see also the test failing on the go mod file. Does it expect a new line at the end of the file?

**Comment by wimspaargaren** [comment]
I signed it!

File: runtime/marshal_httpbodyproto.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+		DefaultMarshaler: &JSONPb{OrigName: true},
+	}
+}
+
+// HTTPBodyMarshaler is a Marshaler which supports marshaling of a
+// google.api.HttpBody message as the full response body if it is
+// the actual message used as the response. If not, then this will
+// simply fallback to the JSONPb marshaler.
+type HTTPBodyMarshaler struct {
+	DefaultMarshaler Marshaler
```
Comment (johanbrandhorst) [inline runtime/marshal_httpbodyproto.go]: If you embed this directly you can avoid defining the methods which just call the underlying implementations:

```go
type HTTPBodyMarshaler struct {
	Marshaler
}

func SetHTTPBodyMarshaler(serveMux *ServeMux) {
	serveMux.marshalers.mimeMap[MIMEWildcard] = &HTTPBodyMarshaler{
		Marshaler: &JSONPb{OrigName: true},
	}
}
```

File: runtime/marshal_httpbodyproto_test.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
+package runtime_test
+
+import (
+	"bytes"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"google.golang.org/genproto/googleapis/api/httpbody"
+	"testing"
```
Comment (johanbrandhorst) [inline runtime/marshal_httpbodyproto_test.go]: Sorry, same thing about the imports please.

**Comment by johanbrandhorst** [comment]
> A couple of comments and I'm also wondering if you could put together something for the documentation as well? Along the lines of the patch examples: https://grpc-ecosystem.github.io/grpc-gateway/docs/patch.html.

What do you think about this?

File: go.mod (hunk 4 above)
Comment (wimspaargaren) [reply go.mod]: The go mod and sum contain a new line at the end of the file. My editor removed these, but I added them again.

Comment (johanbrandhorst) [reply go.mod]: Nice, that would explain it

File: runtime/marshal_httpbodyproto.go (hunk 5 above)
Comment (wimspaargaren) [reply runtime/marshal_httpbodyproto.go]: Adjusted.

File: runtime/marshal_httpbodyproto_test.go (hunk 6 above)
Comment (wimspaargaren) [reply runtime/marshal_httpbodyproto_test.go]: Fixed.

File: runtime/marshal_httpbodyproto.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+import (
+	"io"
+
+	"google.golang.org/genproto/googleapis/api/httpbody"
+)
+
+// SetHTTPBodyMarshaler overwrite  the default marshaler with the HTTPBodyMarshaler
+func SetHTTPBodyMarshaler(serveMux *ServeMux) {
+	serveMux.marshalers.mimeMap[MIMEWildcard] = &HTTPBodyMarshaler{
+		&JSONPb{OrigName: true},
```
Comment (johanbrandhorst) [inline runtime/marshal_httpbodyproto.go]: Please still specify this explicitly:

```
&HTTPBodyMarshaler{
    Marshaler: &JSONPb{OrigName: true},
}
```

File: runtime/marshal_httpbodyproto.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+// NewDecoder returns a Decoder which reads JSON stream from "r".
+func (h *HTTPBodyMarshaler) NewDecoder(r io.Reader) Decoder {
+	return h.Marshaler.NewDecoder(r)
+}
+
+// NewEncoder returns an Encoder which writes JSON stream into "w".
+func (h *HTTPBodyMarshaler) NewEncoder(w io.Writer) Encoder {
+	return h.Marshaler.NewEncoder(w)
+}
```
Comment (johanbrandhorst) [inline runtime/marshal_httpbodyproto.go]: You can remove `NewEncoder`, `NewDecoder` and `Unmarshal` now. The embedded methods will be called automatically.

Comment (wimspaargaren) [reply runtime/marshal_httpbodyproto.go]: Good point :)

File: runtime/marshal_httpbodyproto.go (hunk 7 above)
Comment (wimspaargaren) [reply runtime/marshal_httpbodyproto.go]: Done.

**Comment by johanbrandhorst** [comment]
> > A couple of comments and I'm also wondering if you could put together something for the documentation as well? Along the lines of the patch examples: https://grpc-ecosystem.github.io/grpc-gateway/docs/patch.html.
> 
> What do you think about this?

Bump :D

**Comment by wimspaargaren** [comment]
Not sure if it is added at the right place, but I added a small README for the httpbody :)

File: docs/_docs/httpbody.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+## Example Usage
+1. Create a mux with the HTTP Body Marshaler as option. 
+
+```golang 
+	mux := runtime.NewServeMux(runtime.SetHTTPBodyMarshaler)
+```
+2. Define your service in gRPC with an httpbody response message
+
+```golang
```
Comment (johanbrandhorst) [inline docs/_docs/httpbody.md]: Maybe add the `import "google/api/http.proto";` as well

File: docs/_docs/httpbody.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
+---
+category: documentation
+---
+
+# HttpBody message Feature
+The [HTTP Body](https://github.com/googleapis/googleapis/blob/master/google/api/httpbody.proto) messages allows a response message to be specified with custom data content and a custom content type header.
```
Comment (johanbrandhorst) [inline docs/_docs/httpbody.md]: `The values included in the HTTPBody response will be used verbatim in the returned message from the gateway. Make sure you format your response carefully!`.

File: docs/_docs/httpbody.md [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+service HttpBodyExampleService {
+
+ 	rpc HelloWorld(google.protobuf.Empty) returns (google.api.HttpBody) {
+        option (google.api.http) = {
+            get: "/helloworld"
+        };
+	}	
+
+}
```
Comment (johanbrandhorst) [inline docs/_docs/httpbody.md]: Something happened to the indentation here.

Comment (wimspaargaren) [reply docs/_docs/httpbody.md]: I added the service around the function definition.

Comment (johanbrandhorst) [reply docs/_docs/httpbody.md]: The `option` should be indented more I think

Comment (wimspaargaren) [reply docs/_docs/httpbody.md]: Seems to be fixed now.

Comment (johanbrandhorst) [reply docs/_docs/httpbody.md]: I... are you messing with me :joy:? I think you added one tab too many now.

Comment (johanbrandhorst) [reply docs/_docs/httpbody.md]: Nevermind, much better

**Comment by wimspaargaren** [comment]
No problem!

**Comment by klim0v** [comment]
how can I specify a file name and extension?

**Comment by johanbrandhorst** [comment]
I think that'd have to be part of your proto file annotation:
```
get: /picture.jpg
```

**Pull request merged**